src/Microsoft.VisualStudio.Editors/Common/ClientShimUtils.vb

```diff
@@ -0,0 +1,210 @@
+﻿' Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+Imports System.Runtime.InteropServices
+Imports Microsoft.VisualStudio.OLE.Interop
+Imports Microsoft.VisualStudio.ProjectSystem.Query
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel
+Imports Microsoft.VisualStudio.Shell.Interop
+Imports Microsoft.VisualStudio.Shell
+Imports Microsoft.VisualStudio.Editors.Interop
+Imports System.IO
+
+Namespace Microsoft.VisualStudio.Editors.Common
+
+    ''' <summary>
+    ''' Shim layer to handle nexus and non-nexus APIs based on project
+    ''' </summary>
+    Friend NotInheritable Class ClientShimUtils
+
+        Public Shared Function VSHierarchyShim(itemId As UInteger, vsHierarchy As IVsHierarchy, serviceProvider As System.IServiceProvider) As IVsHierarchy
+
+            If KnownUIContexts.CloudEnvironmentConnectedContext.IsActive = False Then
+                Return vsHierarchy
+            End If
+
+            ' Is Misc File and is nexus client then create the cps object
+            If IsMiscProjectFile(itemId, vsHierarchy) Then
+                Dim path As String = ""
+                vsHierarchy.GetCanonicalName(itemId, path)
+                Return New ClientVsHierarchy(vsHierarchy, path, serviceProvider)
+            End If
+
+            Return vsHierarchy
+        End Function
+
+        Private Shared Function IsMiscProjectFile(itemId As UInteger, vsHierarchy As IVsHierarchy) As Boolean
+
+            Dim pguid As Guid
+            vsHierarchy.GetGuidProperty(itemId, __VSHPROPID.VSHPROPID_ProjectIDGuid, pguid)
```

Comment:
ProjectID is the unique identifier of a project within a solution, you want to get the type of the project, which is provided by TypeGuid.